proxy/src/auth/backend/console.rs

```diff
@@ -5,26 +5,68 @@ use crate::{
     auth::{self, AuthFlow, ClientCredentials},
     compute,
     error::{io_error, UserFacingError},
-    http, scram,
+    http, sasl, scram,
     stream::PqStream,
 };
 use futures::TryFutureExt;
-use serde::{Deserialize, Serialize};
+use reqwest::StatusCode as HttpStatusCode;
+use serde::Deserialize;
 use std::future::Future;
 use thiserror::Error;
 use tokio::io::{AsyncRead, AsyncWrite};
-use tracing::{error, info, info_span};
+use tracing::{error, info, info_span, warn, Instrument};
 
+/// A go-to error message which doesn't leak any detail.
 const REQUEST_FAILED: &str = "Console request failed";
 
+/// Common console API error.
 #[derive(Debug, Error)]
-#[error("{}", REQUEST_FAILED)]
-pub struct TransportError(#[from] std::io::Error);
+pub enum ApiError {
+    /// Error returned by the console itself.
+    #[error("{REQUEST_FAILED} with {}: {}", .status, .text)]
+    Console {
+        status: HttpStatusCode,
+        text: Box<str>,
+    },
+
+    /// Various IO errors like broken pipe or malformed payload.
+    #[error("{REQUEST_FAILED}: {0}")]
+    Transport(#[from] std::io::Error),
+}
+
+impl ApiError {
+    /// Returns HTTP status code if it's the reason for failure.
+    fn http_status_code(&self) -> Option<HttpStatusCode> {
+        use ApiError::*;
+        match self {
+            Console { status, .. } => Some(*status),
+            _ => None,
+        }
+    }
+}
 
-impl UserFacingError for TransportError {}
+impl UserFacingError for ApiError {
+    fn to_string_client(&self) -> String {
+        // Ask @neondatabase/control-plane for review before adding more.
+        let harmless = [
+            HttpStatusCode::LOCKED,                // 423
+            HttpStatusCode::NOT_FOUND,             // 404
+            HttpStatusCode::INTERNAL_SERVER_ERROR, // 500
```

Comment:
I'm still not sure about that one, too.

Reply:
> Can there be arbitrary text in 500 messages?

We shouldn't leak it, but I think we can now. Something like https://github.com/neondatabase/cloud/pull/3195 should fix that.

Generally, the flow is following:
- every generic Go error is rendered as vague 'internal error' or 'unknown internal error' (this depends on API, I think we can make it universal) with code 500 automatically by errors middleware
- also we have semantic errors, which are widely used if we have something to say to the user. They are created by methods like `semerr.WrapWithInvalidInput(err, msg)`, where `msg` is what will be rendered to the client. Each semantic error has a corresponding HTTP code
- yet, there is a small intersection of these two cases, we have also `semerr.WrapWithInternal()`, so that we can render some custom text on 500s too. Usually, this is not needed, but for some flows like OAuth we usually cannot say who is responsible for error, but we can hint that there is a probably some missconfiguration in the flow. I see misuse of this in the code, mostly in the v1, but I think it could be useful

I hope that this clarifies a bit the question about errors.

Reply:
Since https://github.com/neondatabase/cloud/pull/3195 is merged, I guess we're done here. @ololobus concerns?

Reply:
> it'd be ridiculous if there was nothing in the console (idk, defensive programming?) to prevent leaks from happening in user-facing APIs

And this actually brings my previous question back https://github.com/neondatabase/neon/pull/2750#issuecomment-1335501487. Management API that proxy uses for auth isn't something user-facing, obviously, as you get role role secrets out of it.

So I turned detailed errors in it for now as 'defensive programming' measure to do not accidentally leak internal state through proxy. But that was actually an intended behavior (there was even a test for that), since we want to see detailed errors in the management API (it's internal thing after all).

That said, in the long-term run this is a proxy responsibility to decide what to show to the user and what to don't. The same as we receive a raw error from pageserver API in the control-plane, but only save / log it internally, without showing to the user. One option to implement that could be to include both `error` and `message` in the management API response, so that proxy can show the content of `message`, but log `error`

Reply:
TL;DR I don't think that sending everything that we receive through some internal API to the client is a viable long term solution.

Reply:
> That said, in the long-term run this is a proxy responsibility to decide what to show to the user and what to don't. 

Sorry, but I can't agree with that. If we're talking about `{"error": "..."}` alone, proxy can't possibly know severity and importance of this message, because it's completely unconstrained. I'd go as far as to say that we should never implicitly assume that our APIs are private, because that's a very dangerous assumption! Consider this: why do we give an error message if it should be kept a secret? We could just log it in the console and return nothing but the status code; problem solved.

There are very specific cases when the proxy knows better, like when there's no secret, but that's enough of a burden.

Reply:
> Consider this: why do we give an error message if it should be kept a secret? We could just log it in the console and return nothing but the status code; problem solved.

Because when you set up new pageserver or do tenant relocation, you usually want to see the detailed error message in the internal management API response without going through logs. I don't say that leaking secrets in the logs and error responses is the right way to do, that's generally bad.